### PR TITLE
Refactor step types

### PR DESCRIPTION
**Pull request by keevan**
Initial changes to classnames and current structure based on https://github.com/catalyst/moodle-tool_dataflows/issues/145

- [x] classes should either a prefix like 'source_xyz', 'connector, xyz', or nested ones like flow_filter_xyz
- [x] most of these rough groupings of steps should inherit from a generic class which sets the inputs and outputs (e.g. flow_step, connector_step will be base classes for those various step types)
- [x] have methods which say what type or 'group' this step is which will be a language string used to group the steps in the step chooser page
image

### CODE REVIEW

File: classes/step/base_step.php [hunk 1]
```diff
@@ -40,11 +36,19 @@ abstract class base_step {
      */
     protected $component = 'tool_dataflows';
 
-    /** @var int[] number of input streams (min, max) */
-    protected $inputstreams = [0, 1];
+    /* Input and Output flow and connectors, default all things to zero */
```
Comment (brendanheywood) [inline classes/step/base_step.php:39]: As changing namespace breaks all the existing test flows lets also move this to local and get it done so we don't have to change it again

File: templates/dropdown.mustache [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    * label the text or html content of the main button
+    * hastriggers whether or not the there were triggers provided
+    * hasconnectors whether or not the there were connectors provided
+    * hasflows whether or not the there were flows provided
+    * items collection of known categories of step types (trigger, flow, connectors)
+
+    Example context (json):
+    { "status": "OK"}
+}}
+<span class="badge badge-secondary">{{status}}</span>
```
Comment (brendanheywood) [inline templates/dropdown.mustache:38]: This is pretty broken:
image

We are going to end up with lots of step types so whatever popup we have need to be able to scroll internal like a modal.

Maybe the simplest step is just make the chooser its own page and we can make it js later when we get to filtering the chooser based on various constraints (eg I want to add a step that starts with a flow input)

File: templates/dropdown.mustache [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        aria-haspopup="true"
+        aria-expanded="false">
+        {{{label}}}
+    </button>
+    <div class="dropdown-menu" aria-labelledby="tool_dataflows_dropdown">
+        <!-- triggers -->
+        {{#hastriggers}}
+            <h6 class="dropdown-header">{{#str}} stepgrouptriggers, tool_dataflows {{/str}}</h6>
+            {{#items.triggers}}
+            <a class="dropdown-item" href="{{{href}}}">{{label}}</a>
```
Comment (brendanheywood) [inline templates/dropdown.mustache:55]: It doesn't matter that you choose, the selection doesn't end up pin the form

image

File: templates/dropdown.mustache (hunk 2 above)
Comment (keevan) [reply templates/dropdown.mustache:38]: I agree this is odd behaviour. The default bootstrap dropdown actually drops down and doesn't drop... in the middle? It'd be good if there was a dropdown + filter so you can just filter down to the name you want, but another page might work as well.

Comment (keevan) [reply templates/dropdown.mustache:38]: Also _note to self_ to fix up the mustache comments/example.

File: templates/dropdown.mustache (hunk 3 above)
Comment (brendanheywood) [reply templates/dropdown.mustache:55]: For the time being lets split it into a new page

**Comment by keevan** [comment]
Step chooser page:
image

File: remove-step.php [hunk 4]
```diff
@@ -23,7 +23,7 @@
  * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
 
-use tool_dataflows\step;
+use tool_dataflows\local\step;
```
Comment (brendanheywood) [inline remove-step.php:26]: step deletion no longer works:

Exception - Class 'tool_dataflows\local\step' not found

[More information about this error](https://docs.moodle.org/401/en/error/moodle/generalexceptionmessage)

File: run.php [hunk 5]
```diff
@@ -25,7 +25,7 @@
 
 use tool_dataflows\dataflow;
 use tool_dataflows\visualiser;
-use tool_dataflows\execution\engine;
+use tool_dataflows\local\execution\engine;
```
[the remark this answers is not in the document]
Comment (brendanheywood) [reply run.php:28]: This might be a bug specific the cron trigger step and not a generic bug, maybe this can wait

File: remove-step.php (hunk 4 above)
Comment (keevan) [reply remove-step.php]: Fixed + pushed

**Pull request merged**